lib/utils/log/file_writer.go

```diff
@@ -0,0 +1,170 @@
+/*
+ * Teleport
+ * Copyright (C) 2024  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package log
+
+import (
+	"context"
+	"io/fs"
+	"log/slog"
+	"os"
+	"path/filepath"
+	"sync"
+	"sync/atomic"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/gravitational/trace"
+)
+
+var (
+	// fileSharedWriter global file shared writer with the ability to reopen the log file
+	// in the same mode and with the same flags, either by function call or by filesystem
+	// notification subscription on the changes.
+	fileSharedWriter *FileSharedWriter
+
+	// mu protects global setter/getter of the file shared writer.
+	mu sync.Mutex
+)
+
+// FileSharedWriter is similar to SharedWriter except that it requires a `os.File` instead of a `io.Writer`.
+// This is to allow the File reopen required by logrotate and similar tools.
+type FileSharedWriter struct {
+	logFile  atomic.Pointer[os.File]
+	fileFlag int
+	fileMode fs.FileMode
+	watcher  *fsnotify.Watcher
+
+	lock sync.Mutex
+}
+
+// SetFileSharedWriter sets global file shared writer.
```

Comment:
Something like this?

```suggestion
// SetFileSharedWriter sets the global file shared writer, closing the current
// one (if any). The returned error is the error from the closing operation.
```